sdks/python/apache_beam/io/external/jdbc.py

```diff
@@ -0,0 +1,116 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+  PTransforms for supporting Jdbc in Python pipelines. These transforms do not
+  run a Jdbc client in Python. Instead, they expand to ExternalTransforms
+  which the Expansion Service resolves to the Java SDK's JdbcIO.
+
+  Note: To use these transforms, you need to start a Java Expansion Service.
+  Please refer to the portability documentation on how to do that. Flink Users
+  can use the built-in Expansion Service of the Flink Runner's Job Server. The
+  expansion service address has to be provided when instantiating the
+  transforms.
+
+  If you start Flink's Job Server, the expansion service will be started on
+  port 8097. This is also the configured default for this transform. For a
```

Comment:
It looks to me like the default is to start an expansion service with `sdks:java:io:expansion-service:shadowJar`, not to use port 8097, is this out of date?

Reply:
There are 2 approaches.
1. Start flink job server using `sdks:java:io:expansion-service:runShadow` which runs expansion server at localhost:8097
2. Create shadow-jars for expansion service and flink job server via gradle `sdks:java:io:expansion-service:shadowJar`, `runners:flink:1.10:job-server:shadowJar` and leave expansion_service as None, then those jars will be used with random ports.
I din't dig into it but in 2) at first it tries to run expansion service and when it fails it runs the flink job server.

My main example for the code I wrote was kafka.py and generate_sequence external transforms

Reply:
You can run the flink job server with ```runners:flink:1.10:job-server:runShadow``` - the expansion server will be available at port 8097.

On the other hand when you leave expansion_service as None, then ```sdks:java:io:expansion-service:shadowJar``` will be taken with a random port. But I found that flink job server's jar is still needed for it to work.

Reply:
But where does the code for (1) live on the python side?

As far as I can tell the only two options for this transform are either overriding the expansion service with the `expansion_service=` parameter, or the default - running shadowJar with `BeamJarExpansionService` (what you have in (2)). I can't find any logic that looks for an expansion service on port 8097.  Am I missing something?

Reply:
On the python side you can add a parameter expansion_service='localhost:8097' and it will use this address instead of the one from BeamJarExpansionService.
In most cases a user of the transform will start a container with flink job server which exposes localhost:8097 and use this address.

Reply:
Ok, but my concern is that this docstring indicates using the expansion service at localhost:8097 is the "configured default for this transform," which doesn't seem to be the case. It looks to me like the default is to use `BeamJarExpansionService('sdks:java:io:expansion-service:shadowJar')`, which will start up an expansion service in a subprocess. 

I think that's a great default and should be used in the vast majority of cases, I don't think there's any reason to build JdbcIO into the flink expansion service and encourage users to reference that. My understanding is the flink expansion service was just a temporary solution before we had the ability to run expansion services with `BeamJarExpansionService`, so there's no need to use it anymore (@chamikaramj, @robertwb to confirm or correct me).

I think you should:
- [ ] Update the docstring to indicate the default is to use `sdks:java:io:expansion-service:shadowJar`, or even not reference it at all since it should happen transparently for the user.
- [ ] Drop mention of the Flink job server and it's expansion service.

Reply:
I can see now where the misunderstanding came from - I was certain that the default is to run ```docker run --net=host apache/beam_flink1.10_job_server:latest``` as is in flink runner docs for portable setup. But it limited the user to FlinkRunner which in the future will be not sufficient.

Really cool that now ```BeamJarExpansionService``` is the default.

Reply:
I've removed most of those information as when the default is to use BeamJarExpansionService(sdks:java:io:expansion-service:shadowJar) they are redundant.

Reply:
We updated Kafka to build or download the IO expansion jar by default. Probably you can follow the same model here. I updated documentation for Kafka recently: https://github.com/apache/beam/pull/11928/files

BTW one downside with a single IO expansion jar is that the size of the jar increases linearly with the number of cross-language wrappers that we have. This will increase the pipeline setup time for users since a larger jar has to be staged. What is the size increase of IO expansion jar due to this wrapper ? 

cc: @robertwb

Reply:
The Jar is about 1Mb larger

Reply:
Staging time is a valid concern; thanks for calling this out. I think 1MB is a small price to pay for the convenience here--we could have dozens of wrappers here of that size without meaningfully affecting staging time. If it were 10s of MB, or we were getting into a long tail of obscure IOs, that starts to be a different story.

Reply:
Yeah, 1 MB is fine. I think we should continue to monitor the size of this dependency in future Python wrappers that add to it though to make sure that the size does not get increased significantly.